### ISSUE TITLE
Use Span<T> in ArrayBuffer

### ISSUE THREAD

**Issue by Quppa** [maxmind/MaxMind-DB-Reader-dotnet#45]
The `ArrayBuffer` class is a good candidate for the upcoming [`System.Span<T>`](https://msdn.microsoft.com/en-us/magazine/mt814808.aspx) class (replacing the existing method that makes copies of parts of the byte array).

Once [`System.Memory`](https://www.nuget.org/packages/System.Memory/) is stable I can prepare a PR with these changes.

**Comment by oschwald** [comment]
I agree. I have been looking forward to Span as it should help unnecessary allocations.

**Comment by liguobao** [comment]
I agree.

**Comment by oschwald** [comment]
I looked into this more and I don't think there is much reason to use `Span<T>` here. In https://github.com/maxmind/MaxMind-DB-Reader-dotnet/pull/72, I replaced a couple of places where it would have made sense with code that doesn't allocate a `byte[]`.

The remaining uses of `Read()` are for the `bytes`, `uint128`, and `float` (for .Net Standard 2.0 and .NET Framework). None of the current MaxMind databases use these types so these potential allocations are inconsequential.

Given this, I am going to close this issue. I think it would make sense to revisit if `Span<T>` support is every added for memory mapped files as we could simplify some code and possibly remove the `unsafe` use introduced in the above PR.
